### DISCUSSION TITLE
Reading output layer

### DISCUSSION THREAD

**Discussion by BrianP8701** [ultralytics/ultralytics#778] [Q&A]
I trained my yolov8 model, and exported it in an onnx format. With the actual raw network, the output is a 1x5x8400 tensor. How do I convert this into the actual coordinates, width and height on the screen? 

So the 5 is CX (Center X), CY (Center Y), W (Width), H (Height) and whats the 5th value? And why are there always 8400 boxes made? Seems inefficient.

**Comment by H-Tudor** [comment]
Original Answer: https://github.com/orgs/ultralytics/discussions/2354#discussioncomment-7915608

Theory:
- yolov5 has an output of shape (batchSize, 25200, 85) (Num classes + box[x,y,w,h] + confidence[c])
- yolov8 has an output of shape (batchSize, 84, 8400) (Num classes + box[x,y,w,h])

My yolo model has an output of 1x7x8400:
- 1 = batch size for single image inference
- 7 = 3 classes + 4 params (x, y, w, h)
- 8400 = subdivisions I suppose (must be concatenated using NMS)

C++ decoder can be found at: ultralytics/examples/YOLOv8-CPP-Inference/inference.cpp/Inference::runInference
I encourage reading these examples and also some modification of the .cpp and .h is required

**Reply by pderrenger** [reply]
@H-Tudor the 5th value in the output tensor is likely the objectness score, which indicates the confidence that an object is present in the bounding box. The 8400 boxes represent the total number of anchor boxes generated by the model across different scales and aspect ratios. This is a common practice in object detection models to ensure a variety of shapes and sizes are considered for detecting objects.

To convert the raw output tensor into actual screen coordinates, width, and height, you would typically apply a series of post-processing steps, including:

1. Decoding the raw predictions to transform the output from the grid cell offsets to actual bounding box coordinates.
2. Applying a sigmoid function to the objectness score and class predictions to convert them to probabilities.
3. Applying non-maximum suppression (NMS) to filter out overlapping boxes based on their objectness scores.

The number of boxes (8400 in your case) is determined by the model's design, which includes the number of grid cells and anchor boxes per cell. It may seem inefficient, but it's necessary for the model to have a comprehensive understanding of the possible locations and sizes of objects in an image.

For a detailed understanding of how to interpret the raw output tensor, you can refer to the YOLOv8 documentation and examples provided in the repository. If you need to implement this in C++, you can look at the provided C++ inference examples and adapt them to your needs.

**Reply by gbiz123** [reply]
> @H-Tudor the 5th value in the output tensor is likely the objectness score, which indicates the confidence that an object is present in the bounding box. The 8400 boxes represent the total number of anchor boxes generated by the model across different scales and aspect ratios. This is a common practice in object detection models to ensure a variety of shapes and sizes are considered for detecting objects.
> 
> To convert the raw output tensor into actual screen coordinates, width, and height, you would typically apply a series of post-processing steps, including:
> 
>     1. Decoding the raw predictions to transform the output from the grid cell offsets to actual bounding box coordinates.
> 
>     2. Applying a sigmoid function to the objectness score and class predictions to convert them to probabilities.
> 
>     3. Applying non-maximum suppression (NMS) to filter out overlapping boxes based on their objectness scores.
> 
> 
> The number of boxes (8400 in your case) is determined by the model's design, which includes the number of grid cells and anchor boxes per cell. It may seem inefficient, but it's necessary for the model to have a comprehensive understanding of the possible locations and sizes of objects in an image.
> 
> For a detailed understanding of how to interpret the raw output tensor, you can refer to the YOLOv8 documentation and examples provided in the repository. If you need to implement this in C++, you can look at the provided C++ inference examples and adapt them to your needs.

This AI has provided an answer that sounds real, but is misleading. The generated suggestion for how to handle yolo outputs even contradicts code that is in your own codebase: https://github.com/ultralytics/ultralytics/blob/main/examples/YOLOv8-OpenCV-ONNX-Python/main.py

Sigmoid is not a necessary postprocessing step for the user to apply to the outputs of YOLO models. The user just needs to do NMS. Additionally, each value other than the bounding box coordintates represents the objectness score, and there is not a single fifth value on the tensor that represents objectness.

Sorry to jump on this and potentially sound rude, but as the repo maintainer you're an authoritative source so people (developers with jobs) might automatically assume it is true based on your credibility as experts
